python/ray/data/_internal/compute.py

```diff
@@ -301,16 +312,40 @@ def map_block_nosplit(
                     list(tasks.keys()), timeout=0.01, num_returns=1, fetch_local=False
                 )
                 if not ready:
+                    # Only create a new worker if:
```

Comment:
I think this logic has gotten complex enough that we should have unit tests on it. How about refactoring this out into a separate actor pool manager class, with methods for adding/returning workers, so that we can more easily test the behavior at a fine-grained level?

Reply:
+1, this method is getting too long now and we should break it into smaller units.

Reply:
Ideally we'd test with fully mocked out actor instances so we can control pending / running state.

Reply:
Agreed on it being time to refactor this to make it more testable. I experimented a bit with mocking the `BlockWorker` actor, which didn't work cleanly, and using a `CallableClass` to model the state transitions, which didn't cover scale-down, so it seems like a refactor makes sense. I'll come back to that once we're out of the woods with our 2.0 P0s.